File: scripts/spikes/blake3_conformance/conformance.scala
// P0 spike: Scala side of cross-language audit-hash conformance. Implements the SAME
// frozen byte-grammar as pipeline/shared/audit/audit_chain.py. Run: scala conformance.scala
// Separators built from char codes so the source stays plain ASCII (no control chars).
import java.security.MessageDigest
import java.util.Locale

val US = 0x1f.toChar.toString // unit separator between fields
val RS = 0x1e.toChar.toString // record separator between list elements
val NUL = 0x00.toChar.toString // null sentinel

def cf(v: Any): String = v match {
  case null       => NUL
  case b: Boolean => if (b) "1" else "0"
  case i: Int     => i.toString
  case l: Long    => l.toString
  case d: Double  => String.format(Locale.US, "%.8f", Double.box(d))
  case s: String  => s
  case xs: Seq[_] => xs.map(cf).mkString(RS)
  case other      => throw new RuntimeException("non-canonicalizable: " + other)
}

def canon(fields: Seq[Any]): Array[Byte] =
  fields.map(cf).mkString(US).getBytes("UTF-8")

def hex(bytes: Array[Byte]): String = bytes.map(x => f"${x & 0xff}%02x").mkString

val sample: Seq[Any] = Seq(
  "SPOOFING", "trader-7", "RY", 1718500000000L, 1718500060000L, 4, 0.875, null,
  Seq("e1", "e2", "e3"), Seq("UMIR-2.2"), "equity", 1718500061000L
)

@main def run(): Unit = {
  val b = canon(sample)
  println("BYTES_HEX=" + hex(b))
  println("SHA256=" + hex(MessageDigest.getInstance("SHA-256").digest(b)))
}

File: scripts/spikes/blake3_conformance/conformance.mjs
// P0 spike: JS/TS side of cross-language audit-hash conformance. Same frozen byte-grammar
// as pipeline/shared/audit/audit_chain.py. Run: node conformance.mjs
// Separators from char codes so the source stays plain ASCII.
import crypto from "node:crypto";

const US = String.fromCharCode(0x1f); // unit separator between fields
const RS = String.fromCharCode(0x1e); // record separator between list elements
const NUL = String.fromCharCode(0x00); // null sentinel

function cf(v) {
  if (v === null) return NUL;
  if (typeof v === "boolean") return v ? "1" : "0";
  if (typeof v === "bigint") return v.toString();
  if (typeof v === "number") return Number.isInteger(v) ? v.toString() : v.toFixed(8);
  if (typeof v === "string") return v;
  if (Array.isArray(v)) return v.map(cf).join(RS);
  throw new Error("non-canonicalizable: " + v);
}

function canon(fields) {
  return Buffer.from(fields.map(cf).join(US), "utf-8");
}

const sample = [
  "SPOOFING", "trader-7", "RY", 1718500000000, 1718500060000, 4, 0.875, null,
  ["e1", "e2", "e3"], ["UMIR-2.2"], "equity", 1718500061000,
];

const b = canon(sample);
console.log("BYTES_HEX=" + b.toString("hex"));
console.log("SHA256=" + crypto.createHash("sha256").update(b).digest("hex"));
